### ISSUE TITLE
React component wrapped inside custom element (web component)

### ISSUE THREAD

**Issue by pbenschop** [MatthewHerbst/react-to-print#176]
Thank you for react-to-print. Excellent job! In my application users can toggle between a dark and a light (Bootstrap-based) theme. For when the dark theme is active, I figure I "wrap" my React component inside a custom element (web component) to take advantage of the shadow DOM. That would allow me to switch to light theme just for printing purposes without affecting the dark theme that the user is looking at.

When I use react-to-print to print this React class component directly it works as advertised. When I wrap it inside a custom element, everything seem to go through the motions but nothing is rendered. The react-to-print callbacks are called and the React component's render method is called. When using the custom element directly inside another React component (no printing), it renders to the screen just fine.

Would you have any idea what might be going on here?

**Comment by MatthewHerbst** [comment]
Hello, thanks for sharing your use case. Could you please provide a code example, or more ideally, a working code sandbox? It's hard to understand the problem otherwise.

**Comment by pbenschop** [comment]
Hi Matthew, thanks for looking into this. I reworked your sample code and wrapped ComponentToPrint in a Web Component: [Code Example on CodeSandBox.io](https://codesandbox.io/s/mutable-shape-ejg5l?fontsize=14).

**Comment by MatthewHerbst** [comment]
Thanks! That's really interesting, I'll try and take a look at it this weekend

**Comment by MatthewHerbst** [comment]
I've been looking into this. I don't think the issue has anything to do with the custom HTML element, or with this library, since the custom element prints fine.

It seems that the `#shadow-root (open)` that the element renders is not being printed. I'm honestly not sure I would expect it to.  I think what you'll want to do is either temporarily change the actual DOM (such as by using `onBeforeGetContent`/`onAfterPrint` or by forwarding the `ref` down to the shadowRoot `span`.

Happy to discuss this further, but I think a better venue for the question would be StackOverflow where there is likely more expertise on using React and the ShadowDOM API together.

**Comment by MatthewHerbst** [comment]
(I didn't know about the ShadowDOM API prior to looking at this, so thanks for showing me something new!)

**Comment by pbenschop** [comment]
Thanks so much for spending time to look into this, Matthew. I will continue to explore this and let you know if I end up with a solution.

**Comment by MatthewHerbst** [comment]
Please let me know what you do to end up getting it working, would love to learn more! I think the idea of passing the `ref` down is promising though might be a little complex.

**Comment by pbenschop** [comment]
I agree that if the ref to the shadowRoot span could be passed into react-to-print, it would work. The ref could be provided to the Web Component as a property. However, it does not seem React.createElement has a provision to specify the ref.

I did manage it to get it to work with a custom modification in react-to-print. First I modified the Web Component's "connectedCallback" method to set an id on the span: 

`this.mountPoint.setAttribute("id", "span_containing_react_component");`

Next, I modified the "handlePrint" method of react-to-print:

```
let contentNodes;
if (contentEl.shadowRoot) {
  const reactElement = contentEl.shadowRoot.getElementById('span_containing_react_component');
  contentNodes = findDOMNode(reactElement);
} else {
  contentNodes = findDOMNode(contentEl);
}
```
Basically the if statement determines whether react-to-print has been passed a Web Component or a React component. I am not sure how to best generalize this.

**Comment by MatthewHerbst** [comment]
I think there is an important distinction to make when talking about this: the issue is that the shadowDOM is not being printed. The Web Component is being printed, but it has no content except for the shadowDOM component.

On that note, if it would be helpful and could be made generic, we could add a prop to allow custom selection of the parent element (so, you would pass a function down which when given the root node could then return anything from that).

**Comment by pbenschop** [comment]
Yes, you are right, the shadowDOM is not being printed. What you are proposing, the function prop to return the actual element to be printed given a particular root node, would be an extensible and generic solution that will definitely resolve the issue for us.

**Comment by el2king** [comment]
> I agree that if the ref to the shadowRoot span could be passed into react-to-print, it would work. The ref could be provided to the Web Component as a property. However, it does not seem React.createElement has a provision to specify the ref.
> 
> I did manage it to get it to work with a custom modification in react-to-print. First I modified the Web Component's "connectedCallback" method to set an id on the span:
> 
> `this.mountPoint.setAttribute("id", "span_containing_react_component");`
> 
> Next, I modified the "handlePrint" method of react-to-print:
> 
> ```
> let contentNodes;
> if (contentEl.shadowRoot) {
>   const reactElement = contentEl.shadowRoot.getElementById('span_containing_react_component');
>   contentNodes = findDOMNode(reactElement);
> } else {
>   contentNodes = findDOMNode(contentEl);
> }
> ```
> 
> Basically the if statement determines whether react-to-print has been passed a Web Component or a React component. I am not sure how to best generalize this.

Is there a clear full example (not just snippets) where an element wrapped around `ShadowDOM` can be printed with ReactToPrint?  I'm trying to understand what changes need to be made to support printing in a `ShadowDOM` element, but I don't see where exactly to support this ability, based on the snippets above.

**Comment by el2king** [comment]
> > I agree that if the ref to the shadowRoot span could be passed into react-to-print, it would work. The ref could be provided to the Web Component as a property. However, it does not seem React.createElement has a provision to specify the ref.
> > I did manage it to get it to work with a custom modification in react-to-print. First I modified the Web Component's "connectedCallback" method to set an id on the span:
> > `this.mountPoint.setAttribute("id", "span_containing_react_component");`
> > Next, I modified the "handlePrint" method of react-to-print:
> > ```
> > let contentNodes;
> > if (contentEl.shadowRoot) {
> >   const reactElement = contentEl.shadowRoot.getElementById('span_containing_react_component');
> >   contentNodes = findDOMNode(reactElement);
> > } else {
> >   contentNodes = findDOMNode(contentEl);
> > }
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Basically the if statement determines whether react-to-print has been passed a Web Component or a React component. I am not sure how to best generalize this.
> 
> Is there a clear full example (not just snippets) where an element wrapped around `ShadowDOM` can be printed with ReactToPrint? I'm trying to understand what changes need to be made to support printing in a `ShadowDOM` element, but I don't see where exactly to support this ability, based on the snippets above.

^^^ Would like some thoughts on this.  I see that the `src/index.tsx` file has the code for the default `handlePrint` method.  But with my code, I am using `ReactToPrint` as a dependency, so I don't have access to adjust the source code directly.  What could I do with the `handlePrint` method to account for the shadowDOM issue, when I'm using the ReactToPrint library and not the source code?

**Comment by MatthewHerbst** [comment]
@el2king you could potentially write your own print method (see the `print` prop) that can reference the ShadowDOM you need. I'd like to add generic support for this in the near future, I just haven't had time to get around to it. I'll re-open this issue so it doesn't drop-off my radar. I also always welcome PRs.

**Comment by MatthewHerbst** [comment]
Hello. I'm not sure what I did to fix this but I just checked the Codesandbox example using version `2.14.3` (the latest as of today) and it seems to work, while when using version `2.5.0` which was the version when this issue was opened, it does not work. Please let me know if you continue to see any issues regarding this, but for now I'm going to mark this as closed. Happy holidays/happy new years!

**Comment by stavan-vc** [comment]
I'm facing the same issue using version `2.15.1`. I've some custom elements from the [Shoelace UI](https://shoelace.style/components/). These components are not rendered in the PDF.

**Comment by MatthewHerbst** [comment]
@stavan-vc could you please provide a codesandbox or similar showing the issue? I'm not familiar with Shoelace, and need some way to reproduce whatever you are seeing. It might be worth creating a new issue at that point since this one is pretty old.

**Comment by stavan-vc** [comment]
Here is the [sandbox](https://codesandbox.io/p/sandbox/react-to-print-shoelace-w6hfgf?file=%2Fsrc%2FApp.js%3A24%2C15) using Shoelace UI. If this helps you to reproduce the issue. I also found that #638 also relates here.

**Comment by boehlke** [comment]
This is my solution. It is important to copy the shadow root content into the iframe after the iframe was created. Didn't work otherwise....
```ts
    const onPrint = useReactToPrint({
        onBeforePrint: () => {
            const printWindow = document.getElementById('printWindow') as HTMLIFrameElement;
            const iframeDoc = printWindow.contentDocument!!;
            const sourceElement = contentDivRef.current!!;
            const rootElement = iframeDoc.getElementById(sourceElement.id)!!
            cloneShadowRoots(sourceElement, rootElement)
        },
        content: () => contentDivRef.current!!
    })
```

This is the gist where you can find the clone function; https://gist.github.com/boehlke/a4788addadb1853eba2e4b4c72e77fbe

**Comment by stavan-vc** [comment]
Thanks for the solution @boehlke.

**Comment by MatthewHerbst** [comment]
Thanks for posting @boehlke! I'm working on v3 of this lib with React 19 support, but after that I'd like to come back to some of these other issues. This is probably something that the library should handle for you, so this will be a good guide for me when I do that!

**Comment by boehlke** [comment]
> Thanks for posting @boehlke! I'm working on v3 of this lib with React 19 support, but after that I'd like to come back to some of these other issues. This is probably something that the library should handle for you, so this will be a good guide for me when I do that!

Thank you for this nice library. If you like, I could create a PR.

**Comment by MatthewHerbst** [comment]
PRs are _always_ welcome!

**Comment by boehlke** [comment]
I created the PR, please let me know how I can improve/change it so you might merge it.

**Comment by MatthewHerbst** [comment]
> I created the PR, please let me know how I can improve/change it so you might merge it.

I was traveling the last few days apologies, just commented on it, thank you for making it!
